### DISCUSSION TITLE
OTD Install/Portable Experience

### DISCUSSION THREAD

**Discussion by X9VoiD** [OpenTabletDriver/OpenTabletDriver#649] [Ideas]
# Idea

Semi-pseudo code
```cs
namespace OpenTabletDriver.Desktop.Binary
{
    private abstract BinaryManager
    {
        public BinaryManager()
        {
            // move DesktopPluginManager.Clean() here
        }

        public abstract bool IsFirstRun { get; }

        public abstract bool IsInstalled { get; }

        public bool IsUpdated =>
        {
            if (IsInstalled && Settings.Version != CurrentAssembly.Version && !DebugBuild)
            {
                Settings.Version = CurrentAssembly.Version;
                // Serialize Settings
                return true;
            }
            return false;
        };

        public virtual bool UpdateAvailable => CheckForUpdate();

        public virtual void Install(DirectoryInfo binaryDir)
        {
            // copy binaryDir to install directory
        }

        public virtual void Uninstall(DirectoryInfo binaryDir)
        {
            // move binaryDir to Trash
        }
    }
}
```

Different platforms will extend `BinaryManager`

---

## Windows First Run

Premise:

- User will download a release binary: `OpenTabletDriver`
- User probably did not uninstall other tablet drivers

To determine if this is the very first run, check for existence of `HKCU\SOFTWARE\OpenTabletDriver` registry or `./userdata` folder.

On first run, prompt if user wants to install OTD. (Prompt shows before single instance code runs)

If yes:
- Notify that the user should uninstall other tablet driver for optimal experience
- Copy current OTD into `%localappdata%\OpenTabletDriver\bin`
- Save registry stuff
- Add shortcut
- Launch `%localappdata%\OpenTabletDriver\bin\OpenTabletDriver.UX.Wpf.exe`
- `Environment.Exit(0)`

If no:
- Prompt if the user wants to run OTD in full portable mode
- If full portable, create `./userdata`
- Continue

---

## Update

Update check runs on otd boot after install check, or through app menu.

### Windows

- Download new release, extract
- `BinaryManager.Uninstall(installDir)`, and then
- `BinaryManager.Install(updatedBinaryDir)`

### Linux

- If running portable, we will be able to do the same thing as Windows

### MacOS

> Not sure how MacOS works, but if moving executables are fine and we have write access to where OTD is installed then doing the same thing as Windows will be fine.

# Other Stuff

StartupGreeterWindow will show if `IsFirstRun` or `IsUpdated` is true

**Comment by InfinityGhost** [comment]
Ideally I want to switch to testable IO, like [System.IO.Abstractions](https://github.com/System-IO-Abstractions/System.IO.Abstractions) to perform a lot of this. Ideally we wouldn't have an unnecessary property for a function that already exists as well.

**Comment by InfinityGhost** [comment]
> Not sure how MacOS works, but if moving executables are fine and we have write access to where OTD is installed then doing the same thing as Windows will be fine.

MacOS contains the binaries inside of its OpenTabletDriver.app directory, this is definitely doable.
